File: majorProject/src/main/java/com/aocc/majorproject/ui/SpatialFocusNavigator.java
package com.aocc.majorproject.ui;

import com.aocc.majorproject.input.GamepadInput;

import java.util.List;

/**
 * Cone-weighted spatial focus navigation (similar to Android TV / Unity UI focus).
 * From the current item, picks the nearest focusable in the requested direction
 * whose center lies inside a forward-facing cone.
 */
public final class SpatialFocusNavigator {

    /** Minimum alignment with the search direction (cos 60° → 60° half-angle cone). */
    private static final float MIN_DIRECTION_COS = 0.5f;

    public enum Direction {
        UP, DOWN, LEFT, RIGHT
    }

    private SpatialFocusNavigator() {
    }

    public static Direction directionFrom(GamepadInput.Action action) {
        return switch (action) {
            case UP -> Direction.UP;
            case DOWN -> Direction.DOWN;
            case LEFT -> Direction.LEFT;
            case RIGHT -> Direction.RIGHT;
            default -> null;
        };
    }

    /**
     * @return index of the next focus item, or {@code currentIndex} if none in that direction
     */
    public static int findNext(int currentIndex, Direction direction, List<UiBounds> items) {
        if (items == null || items.isEmpty()) {
            return currentIndex;
        }
        currentIndex = clamp(currentIndex, items.size());
        UiBounds current = items.get(currentIndex);
        float originX = current.centerX();
        float originY = current.centerY();
        float dirX = directionVectorX(direction);
        float dirY = directionVectorY(direction);

        int bestIndex = currentIndex;
        float bestScore = Float.MAX_VALUE;

        for (int i = 0; i < items.size(); i++) {
            if (i == currentIndex) {
                continue;
            }
            UiBounds candidate = items.get(i);
            float dx = candidate.centerX() - originX;
            float dy = candidate.centerY() - originY;
            float distSq = dx * dx + dy * dy;
            if (distSq < 1f) {
                continue;
            }

            float dist = (float) Math.sqrt(distSq);
            float alignment = (dx * dirX + dy * dirY) / dist;
            if (alignment < MIN_DIRECTION_COS) {
                continue;
            }

            float along = dx * dirX + dy * dirY;
            float perpSq = distSq - along * along;
            float score = along * along * 100f + perpSq;

            if (score < bestScore) {
                bestScore = score;
                bestIndex = i;
            }
        }
        return bestIndex;
    }

    private static float directionVectorX(Direction direction) {
        return switch (direction) {
            case LEFT -> -1f;
            case RIGHT -> 1f;
            default -> 0f;
        };
    }

    private static float directionVectorY(Direction direction) {
        return switch (direction) {
            case UP -> -1f;
            case DOWN -> 1f;
            default -> 0f;
        };
    }

    private static int clamp(int index, int size) {
        if (index < 0) {
            return 0;
        }
        if (index >= size) {
            return size - 1;
        }
        return index;
    }
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/ui/SpatialFocusNavigator.kt
package com.aocc.majorproject.ui

import com.aocc.majorproject.input.GamepadInput
import kotlin.math.sqrt

/**
 * Cone-weighted spatial focus navigation (similar to Android TV / Unity UI focus).
 * From the current item, picks the nearest focusable in the requested direction
 * whose center lies inside a forward-facing cone.
 */
object SpatialFocusNavigator {

    /** Minimum alignment with the search direction (cos 60° → 60° half-angle cone). */
    private const val MIN_DIRECTION_COS = 0.5f

    enum class Direction {
        UP, DOWN, LEFT, RIGHT
    }

    @JvmStatic
    fun directionFrom(action: GamepadInput.Action): Direction? {
        return when (action) {
            GamepadInput.Action.UP -> Direction.UP
            GamepadInput.Action.DOWN -> Direction.DOWN
            GamepadInput.Action.LEFT -> Direction.LEFT
            GamepadInput.Action.RIGHT -> Direction.RIGHT
            else -> null
        }
    }

    /**
     * @return index of the next focus item, or `currentIndex` if none in that direction
     */
    @JvmStatic
    fun findNext(currentIndex: Int, direction: Direction, items: List<UiBounds>?): Int {
        if (items.isNullOrEmpty()) {
            return currentIndex
        }
        var index = clamp(currentIndex, items.size)
        val current = items[index]
        val originX = current.centerX().toFloat()
        val originY = current.centerY().toFloat()
        val dirX = directionVectorX(direction)
        val dirY = directionVectorY(direction)

        var bestIndex = index
        var bestScore = Float.MAX_VALUE

        for (i in items.indices) {
            if (i == index) {
                continue
            }
            val candidate = items[i]
            val dx = candidate.centerX() - originX
            val dy = candidate.centerY() - originY
            val distSq = dx * dx + dy * dy
            if (distSq < 1f) {
                continue
            }

            val dist = sqrt(distSq)
            val alignment = (dx * dirX + dy * dirY) / dist
            if (alignment < MIN_DIRECTION_COS) {
                continue
            }

            val along = dx * dirX + dy * dirY
            val perpSq = distSq - along * along
            val score = along * along * 100f + perpSq

            if (score < bestScore) {
                bestScore = score
                bestIndex = i
            }
        }
        return bestIndex
    }

    private fun directionVectorX(direction: Direction): Float {
        return when (direction) {
            Direction.LEFT -> -1f
            Direction.RIGHT -> 1f
            else -> 0f
        }
    }

    private fun directionVectorY(direction: Direction): Float {
        return when (direction) {
            Direction.UP -> -1f
            Direction.DOWN -> 1f
            else -> 0f
        }
    }

    private fun clamp(index: Int, size: Int): Int {
        if (index < 0) {
            return 0
        }
        if (index >= size) {
            return size - 1
        }
        return index
    }
}
